keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -211,63 +309,51 @@ When the value of the discriminator is explicitly changed by the client, it
-    other fields. if more than one field has been added, leave the process so
-    that validation will fail.
-
 #### "At most one" versus "exactly one"
 
-The goal of this proposal is not to change the validation, but to help clients
-to clear other fields in the union. Validation should be implemented for in-tree
-types as it is today, or through "oneOf" properties in CRDs.
+// TODO: check this
+We propose supporting both "at most one" and "exactly one" semantics. 
+
+To distinguish between the two, API authors should tag the discriminator as
+"optional" or "required" based on the desired behavior. We the discriminator is
+set to an empty string (or missing from the object), this will indicate to the
+server to clear all fields of the union (or fail the request if the
+discriminator is required).
 
-In other word, this is proposing to implement "at most one", and the exactly one
-should be provided through another layer of validation (separated).
+### Normalization and Validation
 
-#### Clearing all the fields
+Normalization refers to the process by which the API server attempts to understand
+and correct clients which may provide the server with conflicting or incomplete
+information about a union.
 
-Since the system is trying to do the right thing, it can be hard to "clear
-everything". In that case, each API could decide to have their own "Nothing"
-value in the discriminator, which will automatically trigger a clearing of all
-fields beside "Nothing".
+Issues primarily arise here because of version skew between a client and server,
+such as when a client is unaware of new fields added to a union and thus doesn't
+know how to clear these new fields when trying to set a different.
 
-### Backward compatibilities properties
+For unions with a discriminator (all newly created unions and some existing
+unions), normalization is simple: the server should always respect the
+discriminator.
 
-This normalization process has a few nice properties, especially for dumb
-clients, when it comes to backward compatibility:
-- A dumb client that doesn't know which fields belong to the union can just set
-  a new field and get all the others cleared automatically
-- A dumb client that doesn't know about the discriminator is going to change a
-  field, leave the discriminator as it is, and should still expect the fields to
-  be cleared accordingly
-- A dumb client that knows about the discriminator can change the discriminator
-  without knowing which fields to clear, they will get cleared automatically
+This means two things:
+1. when the server receives a request with a discriminator set to a
+given field, it should clear out any other fields that are set.
+2. when the server receives a request with a discriminator set to a given field,
+   but that given field is empty, it should maintain the value of the field that
+   currently exists.
 
+For union types without a discriminator (only existing unions), normalization is a little more
```

Comment:
Drop these two paragraphs for now, I think we should focus on discriminated unions for now.